File: manual/dcs-apply-card-sdk-demo-oc/dcs-apply-card-sdk-demo-oc/ViewController.m
//
//  ViewController.m
//  dcs-apply-card-sdk-demo-oc
//
//  Created by Ran on 2025/3/18.
//

#import "ViewController.h"
#import <DcsCardSdk/DCSCard.h>

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UITextField *orgCode;
@property (weak, nonatomic) IBOutlet UITextField *orgKey;
@property (weak, nonatomic) IBOutlet UITextField *areaCode;
@property (weak, nonatomic) IBOutlet UITextField *phone;
@property (strong, nonatomic) NSString *authen;
@property (weak, nonatomic) IBOutlet UITextView *logView;
@property (nonatomic, strong) NSMutableString *logText;
@property (weak, nonatomic) IBOutlet UITextField *userId;
@property (weak, nonatomic) IBOutlet UITextField *firstName;
@property (weak, nonatomic) IBOutlet UITextField *lastName;
@property (weak, nonatomic) IBOutlet UITextField *gender;
@property (weak, nonatomic) IBOutlet UITextField *birthDate;
@property (weak, nonatomic) IBOutlet UITextField *nationality;
@property (weak, nonatomic) IBOutlet UITextField *idNo;
@property (weak, nonatomic) IBOutlet UITextField *idType;
@property (weak, nonatomic) IBOutlet UITextField *address;
@property (weak, nonatomic) IBOutlet UITextField *postCode;
@property (weak, nonatomic) IBOutlet UITextField *email;
@property (weak, nonatomic) IBOutlet UITextField *productCode;
@property (weak, nonatomic) IBOutlet UITextField *cardType;
@property (weak, nonatomic) IBOutlet UITextField *cardFlag;
@property (weak, nonatomic) IBOutlet UITextField *partnerID;
@property (weak, nonatomic) IBOutlet UITextField *cardFace;
@property (weak, nonatomic) IBOutlet UITextField *externalWorkflowId;
@property (weak, nonatomic) IBOutlet UITextField *cardToken;
@property (weak, nonatomic) IBOutlet UITextField *environment;
@property (weak, nonatomic) IBOutlet UITextField *custIdToken;

@end

@implementation ViewController

// authorization
- (void)authorization {

    [[DCSCard sharedInstance] authenSdkWithorgCode:self.orgCode.text
                                      orgSecretKey:self.orgKey.text
                                          areaCode:self.areaCode.text
                                       phoneNumber:self.phone.text
                                           bizCode:@"470101"
                                         partnerID:self.partnerID.text
                                    externalUserId:self.userId.text
                                externalWorkflowId:self.externalWorkflowId.text
                                       environment:self.environment.text
                                        completion:^(NSDictionary * _Nonnull responseObject, NSString * _Nonnull error)  {
        // responseObject.code explain
        // 0x00 Submitted successfully
        // 0x01 User exit
        // 0x02 Login Failure
        // 0x03 unknown error
        // 0x04 No authorization information obtained
        // 0x05 Missing parameters
        // 0x06 API call failed
        if (![responseObject[@"data"] isKindOfClass:[NSString class]] && responseObject[@"data"] != nil) {
            NSString *authen = responseObject[@"data"][@"accessToken"];
            if (authen != nil && ![authen isEqualToString:@""]) {
                self.authen = authen;
            }
        }
        if (error != nil) {
            [self log:@{@"error":error}];
        } else {
            [self log:responseObject];
        }
        
    }];
}

// Apply Card
- (void)startApply {
        
    if (self.authen == nil) {
        [self log:@{@"tip":@"Please click on the authorization application to obtain the token first"}];
        return;
    }
    
    [[DCSCard sharedInstance] goApplyforCardbizType:@"D001"
                                            bizCode:@"470102"
                                        accessToken:self.authen
                                          firstName:self.firstName.text
                                           lastName:self.lastName.text
                                             gender:self.gender.text
                                          birthDate:self.birthDate.text
                                        nationality:self.nationality.text
                                             idType:self.idType.text
                                               idNo:self.idNo.text
                                            address:self.address.text
                                           postCode:self.postCode.text
                                              email:self.email.text
                                        productInfo:@[@{@"productCode":self.productCode.text,@"cardType":self.cardType.text,@"cardFace":self.cardFace.text, @"cardFlag":self.cardFlag.text}]
                                          partnerID:self.partnerID.text
                                 externalWorkflowId:self.externalWorkflowId.text
                                        environment:self.environment.text
                                         completion:^(NSDictionary * _Nonnull responseObject, NSString * _Nonnull error) {
        if (error != nil) {
            [self log:@{@"error":error}];
        } else {
            [self log:responseObject];
        }
    }];
}

// Card progress query
- (void)cardProgress {
    
    if (self.authen == nil) {
        [self log:@{@"tip":@"Please click on the authorization application to obtain the token first"}];
        return;
    }
    [[DCSCard sharedInstance] cardProgressInquirybizType:@"D002"
                                                 bizCode:@"470102"
                                             accessToken:self.authen
                                      externalWorkflowId:self.externalWorkflowId.text        environment:self.environment.text
                                           completion:^(NSDictionary * _Nonnull responseObject, NSString * _Nonnull error) {
        
        if (error != nil) {
            [self log:@{@"error":error}];
        } else {
            [self log:responseObject];
        }
    }];
}

// Rest PIN
- (void)goToRestPIN {
    
    if (self.authen == nil) {
        [self log:@{@"tip":@"Please click on the authorization application to obtain the token first"}];
        return;
    }
    
    [[DCSCard sharedInstance] restPinForbizType:@"D007"
                                        bizCode:@"470102"
                                    accessToken:self.authen
                                      cardToken:self.cardToken.text
                                    custIdToken:self.custIdToken.text
                             externalWorkflowId:self.externalWorkflowId.text
                                    environment:self.environment.text
                                     completion:^(NSDictionary * _Nonnull responseObject, NSString * _Nonnull error) {
            
        if (error != nil) {
            [self log:@{@"error":error}];
        } else {
            [self log:responseObject];
        }
        }];
}

- (IBAction)selectTypeClick:(id)sender {
    
    UIAlertController *alertVc = [UIAlertController alertControllerWithTitle:@"please select the type" message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    UIAlertAction *authorization = [UIAlertAction actionWithTitle:@"Authorization" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self authorization];
    }];
    
    UIAlertAction *application = [UIAlertAction actionWithTitle:@"Apply Card" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self startApply];
    }];
    
    UIAlertAction *cardProgress = [UIAlertAction actionWithTitle:@"Card Progress" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self cardProgress];
    }];
    
    UIAlertAction *restPin = [UIAlertAction actionWithTitle:@"Rest Pin" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self goToRestPIN];
    }];
    
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil];
    
    [alertVc addAction:authorization];
    [alertVc addAction:application];
    [alertVc addAction:cardProgress];
    [alertVc addAction:restPin];
    [alertVc addAction:cancel];
    [self presentViewController:alertVc animated:YES completion:nil];
}

- (void)viewDidLoad {
    
    [super viewDidLoad];
    self.logText = [NSMutableString string];
    self.userId.text = [NSString stringWithFormat:@"%@-01",self.orgCode.text];
    [self.orgCode addTarget:self action:@selector(userIdChange:) forControlEvents:UIControlEventEditingChanged];
    [self.view addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(assignKeyboard)]];
}

-(void)userIdChange:(UITextField *)textField{
    self.userId.text = [NSString stringWithFormat:@"%@-01",textField.text];
}

- (void)log:(NSDictionary *)res {
    
    NSString *JSONString;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:res
                                                       options:NSJSONWritingPrettyPrinted
                                                         error:nil];
    if (!jsonData) {
        JSONString = @"Error converting to jsonString";
    } else {
        JSONString  = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    }
    
    [_logText appendString:@"=======================\n"];
    [_logText appendString:JSONString];
    [_logText appendString:@"\n\n"];
    
    [_logView setText:_logText];
    
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        [self.logView setText:self.logText];
        CGFloat originY = self.logView.contentSize.height > self.logView.bounds.size.height ? self.logView.contentSize.height - self.logView.bounds.size.height : 0;
        [self.logView setContentOffset:CGPointMake(0, originY) animated:YES];
    }];
    
}

- (IBAction)clearLog:(id)sender {
    
    [_logText setString:@""];
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        [self.logView setText:self.logText];
        CGFloat originY = self.logView.contentSize.height > self.logView.bounds.size.height ? self.logView.contentSize.height - self.logView.bounds.size.height : 0;
        [self.logView setContentOffset:CGPointMake(0, originY) animated:YES];
    }];
}

- (void)assignKeyboard {
    [self.view endEditing:YES];
}

@end

File: dcs-apply-card-sdk-demo/dcs-apply-card-sdk-demo/ViewController.swift
//
//  ViewController.swift
//  dcs-apply-card-sdk-demo
//
//  Created by Ran on 2025/3/19.
//

import UIKit
import DcsCardSdk

class ViewController: UIViewController {

    @IBOutlet weak var orgCode: UITextField!
    @IBOutlet weak var orgKey: UITextField!
    @IBOutlet weak var areaCode: UITextField!
    @IBOutlet weak var phone: UITextField!
    var authen: String?
    @IBOutlet weak var logView: UITextView!
    var logText = NSMutableString()
    @IBOutlet weak var userId: UITextField!
    @IBOutlet weak var firstName: UITextField!
    @IBOutlet weak var lastName: UITextField!
    @IBOutlet weak var gender: UITextField!
    @IBOutlet weak var birthDate: UITextField!
    @IBOutlet weak var nationality: UITextField!
    @IBOutlet weak var idNo: UITextField!
    @IBOutlet weak var idType: UITextField!
    @IBOutlet weak var address: UITextField!
    @IBOutlet weak var postCode: UITextField!
    @IBOutlet weak var email: UITextField!
    @IBOutlet weak var productCode: UITextField!
    @IBOutlet weak var cardType: UITextField!
    @IBOutlet weak var cardFlag: UITextField!
    @IBOutlet weak var partnerID: UITextField!
    @IBOutlet weak var cardFace: UITextField!
    @IBOutlet weak var externalWorkflowId: UITextField!
    @IBOutlet weak var cardToken: UITextField!
    @IBOutlet weak var environment: UITextField!
    @IBOutlet weak var custIdToken: UITextField!

    override func viewDidLoad() {
        super.viewDidLoad()
        logText = NSMutableString()
        if let orgCode = orgCode.text {
            userId.text = orgCode
        }
        orgCode.addTarget(
            self,
            action: #selector(userIdChange(_:)),
            for: .editingChanged
        )
        let tag = UITapGestureRecognizer(
            target: self,
            action: #selector(assignKeyboard)
        )
        view.addGestureRecognizer(tag)
    }

    @objc
    private func userIdChange(_ textField: UITextField) {
        guard let userId = textField.text else { return }
        self.userId.text = "\(userId)-01"
    }

    @objc
    private func assignKeyboard() {
        view.endEditing(true)
    }

    // MARK: - Authorization
    func authorization() {
        let orgCodeText = orgCode.text ?? ""
        let orgKeyText = orgKey.text ?? ""
        let areaCodeText = areaCode.text ?? ""
        let phoneText = phone.text ?? ""
        let bizCodeText = "470101"
        let partnerIDText = partnerID.text ?? ""
        let externalUserIDText = userId.text ?? ""
        let externalWorkflowIDText = externalWorkflowId.text ?? ""
        let environmentText = environment.text ?? ""

        
        DCSCard.sharedInstance().authenSdkWithorgCode(orgCodeText,
                                                      orgSecretKey: orgKeyText,
                                                      areaCode: areaCodeText,
                                                      phoneNumber: phoneText,
                                                      bizCode: bizCodeText,
                                                      partnerID: partnerIDText,
                                                      externalUserId: externalUserIDText,
                                                      externalWorkflowId: externalWorkflowIDText,
                                                      environment: environmentText) { responseObject, error in
            if let data = responseObject["data"] as? [String: Any],
               let authen = data["accessToken"] as? String, !authen.isEmpty {
                self.authen = authen
            }
            if !error.isEmpty {
                self.log(["error": error])
            } else if let responseDict = responseObject as? [String: Any] {
                self.log(responseDict)
            } else {
                self.log(["error": "Invalid response format"])
            }
        }
    }

    // MARK: - Apply Card
    func startApply() {
        guard let authen = authen else {
            log(["tip": "Please click on the authorization application to obtain the token first"])
            return
        }
        let bizType = "D001"
        let bizCode = "470102"
        let firstNameText = firstName.text ?? ""
        let lastNameText = lastName.text ?? ""
        let genderText = gender.text ?? ""
        let birthDateText = birthDate.text ?? ""
        let nationalityText = nationality.text ?? ""
        let idTypeText = idType.text ?? ""
        let idNoText = idNo.text ?? ""
        let addressText = address.text ?? ""
        let postCodeText = postCode.text ?? ""
        let emailText = email.text ?? ""
        let partnerIDText = partnerID.text ?? ""
        let externalWorkflowIDText = externalWorkflowId.text ?? ""
        let environmentText = environment.text ?? ""
        
        let productCode = productCode.text ?? ""
        let cardType = cardType.text ?? ""
        let cardFace = cardFace.text ?? ""
        let cardFlag = cardFlag.text ?? ""
        let productInfo: [[String: String]] = [
            [
                "productCode": productCode,
                "cardType": cardType,
                "cardFace": cardFace,
                "cardFlag": cardFlag
            ]
        ]
        DCSCard.sharedInstance().goApplyforCardbizType(bizType,
                                                       bizCode: bizCode,
                                                       accessToken: authen,
                                                       firstName: firstNameText,
                                                       lastName: lastNameText,
                                                       gender: genderText,
                                                       birthDate: birthDateText,
                                                       nationality: nationalityText,
                                                       idType: idTypeText,
                                                       idNo: idNoText,
                                                       address: addressText,
                                                       postCode: postCodeText,
                                                       email: emailText,
                                                       productInfo: productInfo,
                                                       partnerID: partnerIDText,
                                                       externalWorkflowId: externalWorkflowIDText,
                                                       environment: environmentText) { responseObject, error in
            if !error.isEmpty {
                self.log(["error": error])
            } else if let responseDict = responseObject as? [String: Any] {
                self.log(responseDict)
            } else {
                self.log(["error": "Invalid response format"])
            }
        }
    }

    // MARK: - Card Progress Query
    func cardProgress() {
        guard let authen = authen else {
            log(["tip": "Please click on the authorization application to obtain the token first"])
            return
        }
        let bizType = "D002"
        let bizCode = "470102"
        let externalWorkflowIDText = externalWorkflowId.text ?? ""
        let environmentText = environment.text ?? ""
        DCSCard.sharedInstance().cardProgressInquirybizType(bizType,
                                                            bizCode: bizCode,
                                                            accessToken: authen,
                                                            externalWorkflowId: externalWorkflowIDText,
                                                            environment: environmentText) { responseObject, error in
            if !error.isEmpty {
                self.log(["error": error])
            } else if let responseDict = responseObject as? [String: Any] {
                self.log(responseDict)
            } else {
                self.log(["error": "Invalid response format"])
            }
        }
    }

    // MARK: - Reset PIN
    func goToRestPIN() {
        guard let authen = authen else {
            log(["tip": "Please click on the authorization application to obtain the token first"])
            return
        }
        let bizType = "D007"
        let bizCode = "470102"
        let cardTokenText = cardToken.text ?? ""
        let custIdTokenText = custIdToken.text ?? ""
        let externalWorkflowIDText = externalWorkflowId.text ?? ""
        let environmentText = environment.text ?? ""
        DCSCard.sharedInstance().restPinForbizType(bizType,
                                                   bizCode: bizCode,
                                                   accessToken: authen,
                                                   cardToken: cardTokenText,
                                                   custIdToken: custIdTokenText,
                                                   externalWorkflowId: externalWorkflowIDText,
                                                   environment: environmentText) { responseObject, error in
            if !error.isEmpty {
                self.log(["error": error])
            } else if let responseDict = responseObject as? [String: Any] {
                self.log(responseDict)
            } else {
                self.log(["error": "Invalid response format"])
            }
        }
    }

    // MARK: - Logging
    func log(_ res: [String: Any]) {
        
        let jsonData = try? JSONSerialization.data(withJSONObject: res, options: .prettyPrinted)
        let jsonString = jsonData.flatMap { String(data: $0, encoding: .utf8) } ?? "Error converting to JSON String"

        logText.append("=======================\n")
        logText.append(jsonString)
        logText.append("\n\n")

        DispatchQueue.main.async {
            self.logView.text = self.logText as String
            let originY = max(0, self.logView.contentSize.height - self.logView.bounds.height)
            self.logView.setContentOffset(CGPoint(x: 0, y: originY), animated: true)
        }
    }

    @IBAction func clearLog(_ sender: Any) {
        logText.setString("")
        DispatchQueue.main.async {
            self.logView.text = self.logText as String
            let originY = max(0, self.logView.contentSize.height - self.logView.bounds.height)
            self.logView.setContentOffset(CGPoint(x: 0, y: originY), animated: true)
        }
    }

    // MARK: - Selection Action
    @IBAction func selectTypeClick(_ sender: Any) {
        let alertVC = UIAlertController(
            title: "Please select the type",
            message: nil,
            preferredStyle: .actionSheet
        )
        alertVC.addAction(UIAlertAction(title: "Authorization", style: .default) { _ in self.authorization() })
        alertVC.addAction(UIAlertAction(title: "Apply Card", style: .default) { _ in self.startApply() })
        alertVC.addAction(UIAlertAction(title: "Card Progress", style: .default) { _ in self.cardProgress() })
        alertVC.addAction(UIAlertAction(title: "Rest Pin", style: .default) { _ in self.goToRestPIN() })
        alertVC.addAction(UIAlertAction(title: "Cancel", style: .cancel))
        present(alertVC, animated: true)
    }
}
